src/client/mod.rs

```diff
@@ -886,7 +887,13 @@ impl Client {
             }
         }
 
-        self.shard_manager_worker.run().await;
+        if let Err(why) = self.shard_manager_worker.run().await {
+            return match why {
+                ShardManagerError::DisallowedGatewayIntents => Err(Error::Gateway(GatewayError::DisallowedGatewayIntents)),
+                ShardManagerError::InvalidGatewayIntents => Err(Error::Gateway(GatewayError::InvalidGatewayIntents)),
+                ShardManagerError::InvalidToken => Err(Error::Gateway(GatewayError::InvalidAuthentication)),
+            }
+        }
```

Comment:
Move the repetitive `Err(Error::Gateway(` from the branches in front of the `match` and let every branch simply return the `GatewayError`.